csrc/id_model/loop_promotion.cpp

```diff
@@ -27,6 +27,65 @@ const ValGraph& LoopPromotionMapBuilder::idGraph(IdMappingMode mode) const {
   return id_model_.idGraph(mode);
 }
 
+std::unordered_map<ValGroup, IterDomain*> LoopPromotionMapBuilder::
+    getProducerPromotionsOfPartiallyInlinedGroups(
+        const std::unordered_map<ValGroup, IterDomain*>&
+            initial_loop_promotion_map,
+        const ValGraph& loop_graph) const {
+  std::unordered_map<ValGroup, IterDomain*> loop_promotion_map_to_propagate;
+
+  for (const auto& map_kv : initial_loop_promotion_map) {
+    const auto& loop_group = map_kv.first;
+    const auto& promotion = map_kv.second;
+
+    // If it's promoted to the exactly mapped domain, should not
+    // need further propagation
+    if (std::all_of(
+            loop_group->begin(),
+            loop_group->end(),
+            [&](Val* loop_group_val) -> bool {
+              return idGraph(IdMappingMode::EXACT)
+                  .disjointValSets()
+                  .strictAreMapped(loop_group_val, promotion);
+            })) {
+      continue;
+    }
+
+    const ExprGroups& uses = loop_graph.getUses(loop_group);
+    if (uses.empty()) {
+      continue;
+    }
+
+    // If this expr group has outputs that are partially inlined, the
+    // number of output groups must be larger than the actual number
+    // of outputs of the expr.
+
+    const int expected_num_consumer_loop_group_count_if_fully_inlined =
+        (int)uses.front()->front()->outputs().size();
+
+    // If there's only one output, it should not cause partial
+    // inline.
+    if (expected_num_consumer_loop_group_count_if_fully_inlined == 1) {
+      continue;
+    }
+
+    ValGroups consumer_loop_groups;
+    for (const ExprGroup& use : loop_graph.getUses(loop_group)) {
+      std::vector<ValGroup> output_loop_groups = loop_graph.outputGroups(use);
+      consumer_loop_groups.pushBack(output_loop_groups);
+    }
+
+    if (consumer_loop_groups.size() ==
+        expected_num_consumer_loop_group_count_if_fully_inlined) {
+      continue;
+    }
```

Comment:
Am I understanding correct that we can `NVF_ERROR(consumer_loop_groups.size() >= expected_num_consumer_loop_group_count_if_fully_inlined)`?

Reply:
I think so. If the two outputs of a split are involved in broadcast forwarding, they could be grouped together, so if that happens, there should be only one output loop group, but I don't think that can happen. Let me insert the assertion.